worker/caasunitprovisioner/worker.go

```diff
@@ -109,10 +110,20 @@ func (p *provisioner) loop() error {
 			for _, appId := range apps {
 				appLife, err := p.config.LifeGetter.Life(appId)
 				if errors.IsNotFound(err) {
+					// Once an application is deleted, remove the k8s service and ingress resources.
+					if err := p.config.ContainerBroker.UnexposeService(appId); err != nil {
+						return errors.Trace(err)
+					}
+					if err := p.config.ContainerBroker.DeleteService(appId); err != nil {
+						return errors.Trace(err)
+					}
 					w, ok := appWorkers[appId]
 					if ok {
+						// Before stopping the application worker, inform it that
+						// the app is gone so it has a chance to clean up.
+						p.appRemoved <- struct{}{}
 						if err := worker.Stop(w); err != nil {
```

Comment:
Shouldn't this be `w.Wait()` rather than `worker.Stop(w)`? Otherwise there's a race between the `catacomb.Dying()` and `appRemoved` branches of the select in the application worker loop - if the catacomb one wins then the cleanup won't be done.

Reply:
Oh no, I see - appRemoved is an unbuffered channel, so you know the signal to clean up the unit has been received. Maybe worth a comment? Or it should still be a wait anyway?

Reply:
Stop() does a wait internally

Reply:
But I've made it so that the application worker acts on any app removal requests prior to shutdown so as to prevent stuff from accidentally being left behind.